sql/core/src/main/scala/org/apache/spark/sql/execution/command/CommandUtils.scala

```diff
@@ -47,15 +48,27 @@ object CommandUtils extends Logging {
     }
   }
 
-  def calculateTotalSize(sessionState: SessionState, catalogTable: CatalogTable): BigInt = {
+  def calculateTotalSize(spark: SparkSession, catalogTable: CatalogTable): BigInt = {
+    val sessionState = spark.sessionState
     if (catalogTable.partitionColumnNames.isEmpty) {
       calculateLocationSize(sessionState, catalogTable.identifier, catalogTable.storage.locationUri)
     } else {
       // Calculate table size as a sum of the visible partitions. See SPARK-21079
       val partitions = sessionState.catalog.listPartitions(catalogTable.identifier)
-      partitions.map { p =>
-        calculateLocationSize(sessionState, catalogTable.identifier, p.storage.locationUri)
-      }.sum
+      val paths = partitions.map(x => new Path(x.storage.locationUri.get))
+      val stagingDir = sessionState.conf.getConfString("hive.exec.stagingdir", ".hive-staging")
+      val pathFilter = new PathFilter with Serializable {
+        override def accept(path: Path): Boolean = {
+          val fileName = path.getName
+          (!fileName.startsWith(stagingDir) &&
+            // Ignore metadata files starting with "_"
+            !fileName.startsWith("_"))
```

Comment:
We already have the function to check if it is a data file, or not. Can we reuse this? https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningAwareFileIndex.scala#L232

Reply:
So we are not creating an instance of `InMemoryFileIndex` class so `isDataPath` is not accessible here. I think it would be simpler to add these checks here.

Reply:
How about just moving the function into somewhere that can be accessed from here?

Reply:
It is a generic method and only used within `PartitioningAwareFileIndex`, so we can move it to `CommandUtils`. Does that sound okay?

Reply:
How about `DataSourceUtils`?

Reply:
Done. 
Also, we are not doing this check when `calculateLocationSize` is called directly. I will file a different PR for this as this is not related to AnalyzeTableCommand.